packages/medusa/src/services/product-variant-inventory.ts

```diff
@@ -433,33 +433,66 @@ class ProductVariantInventoryService extends TransactionBaseService {
         }
       )
 
-    if (reservationCount) {
-      let reservation = reservations[0]
+    reservations.sort((a, _) => {
+      if (a.location_id === locationId) {
+        return -1
+      }
+      return 0
+    })
+
+    if (quantity > 0) {
+      // if quantity is positive, we need to adjust the first reservation with the correct location
+      const firstReservation = reservations[0]
+      if (firstReservation) {
+        await this.inventoryService_
+          .withTransaction(this.activeManager_)
+          .updateReservationItem(firstReservation.id, {
+            quantity: firstReservation.quantity + quantity,
+          })
+      } else {
+        await this.reserveQuantity(variantId, quantity, {
+          lineItemId,
+          locationId,
+        })
 
-      reservation =
-        reservations.find(
-          (r) => r.location_id === locationId && r.quantity >= quantity
-        ) ?? reservation
+        return
+      }
+    }
 
-      const productVariantInventory = await this.retrieve(
-        reservation.inventory_item_id,
-        variantId
-      )
+    if (reservationCount) {
+      const inventoryItems = await this.listByVariant(variantId)
+      const productVariantInventory = inventoryItems[0]
 
-      const reservationQtyUpdate =
-        reservation.quantity +
+      const deltaUpdate = Math.abs(
         quantity * productVariantInventory.required_quantity
+      )
 
-      if (reservationQtyUpdate === 0) {
-        await this.inventoryService_
-          .withTransaction(this.activeManager_)
-          .deleteReservationItem(reservation.id)
-      } else {
+      const exactReservation = reservations.find(
+        (r) => r.quantity === deltaUpdate && r.location_id === locationId
+      )
+      if (exactReservation) {
         await this.inventoryService_
           .withTransaction(this.activeManager_)
-          .updateReservationItem(reservation.id, {
-            quantity: reservationQtyUpdate,
-          })
+          .deleteReservationItem(exactReservation.id)
+        return
+      }
+
+      // take reservations and delete until we reach the quantity
+      let remainingQuantity = deltaUpdate
+      for (const r of reservations) {
```

Comment:
suggestion: 
```ts
let remainingQuantity = deltaUpdate
 
const reservationsToDelete = []
const reservationToUpdate = null
reservations.forEach(reservastion => {
  if (reservastion.quantity <= remainingQuantity) {
    reservationsToDelete.push(reservastion)
  }  else {
    reservationToUpdate = reservastion
  }
})
 
if (reservationsToDelete.length) {
  await this.inventoryService_
      .withTransaction(this.activeManager_)
      .deleteReservationItem(...) // bulk
}

if (reservationToUpdate) {
  await this.inventoryService_
      .withTransaction(this.activeManager_)
      .updateReservationItem(reservationToUpdate.id, {
        quantity: reservationToUpdate.quantity - remainingQuantity,
      })
}
```

It would reduce the number of operations and queries etc etc wdyt?